File: haskell/src/Y2015/Day1.hs
module Y2015.Day1 (solve) where

parseInstruction :: Char -> Int
parseInstruction '(' = 1
parseInstruction ')' = -1

solve :: String -> (String, String)
solve input = (part1, part2)
  where
    instructions = filter (/= '\n') input
    part1 = show . sum $ map parseInstruction instructions
    part2 = show . length . takeWhile (/=(-1)) $ scanl (\acc c -> acc + parseInstruction c) 0 instructions

File: clojure/src/y2015/day1.clj
(ns y2015.day1
  (:require
    [util.collection :refer[index-of]]))

(defn traverse [acc c]
  (+ acc ({ \( 1 \) -1 } c)))

(defn solve [input]
  (let [part1 (reduce traverse 0 (seq input))
        part2 (index-of -1 (reductions traverse 0 input))]
        [part1 part2]))
